// @azure/arm-containerinstance/esm/models/mappers.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
import * as tslib_1 from "tslib";
import { CloudErrorMapper, BaseResourceMapper } from "@azure/ms-rest-azure-js";
export var CloudError = CloudErrorMapper;
export var BaseResource = BaseResourceMapper;
export var ContainerPort = {
    serializedName: "ContainerPort",
    type: {
        name: "Composite",
        className: "ContainerPort",
        modelProperties: {
            protocol: {
                serializedName: "protocol",
                type: {
                    name: "String"
                }
            },
            port: {
                required: true,
                serializedName: "port",
                type: {
                    name: "Number"
                }
            }
        }
    }
};
export var EnvironmentVariable = {
    serializedName: "EnvironmentVariable",
    type: {
        name: "Composite",
        className: "EnvironmentVariable",
        modelProperties: {
            name: {
                required: true,
                serializedName: "name",
                type: {
                    name: "String"
                }
            },
            value: {
                serializedName: "value",
                type: {
                    name: "String"
                }
            },
            secureValue: {
                serializedName: "secureValue",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var ContainerState = {
    serializedName: "ContainerState",
    type: {
        name: "Composite",
        className: "ContainerState",
        modelProperties: {
            state: {
                serializedName: "state",
                type: {
                    name: "String"
                }
            },
            startTime: {
                serializedName: "startTime",
                type: {
                    name: "DateTime"
                }
            },
            exitCode: {
                serializedName: "exitCode",
                type: {
                    name: "Number"
                }
            },
            finishTime: {
                serializedName: "finishTime",
                type: {
                    name: "DateTime"
                }
            },
            detailStatus: {
                serializedName: "detailStatus",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var Event = {
    serializedName: "Event",
    type: {
        name: "Composite",
        className: "Event",
        modelProperties: {
            count: {
                serializedName: "count",
                type: {
                    name: "Number"
                }
            },
            firstTimestamp: {
                serializedName: "firstTimestamp",
                type: {
                    name: "DateTime"
                }
            },
            lastTimestamp: {
                serializedName: "lastTimestamp",
                type: {
                    name: "DateTime"
                }
            },
            name: {
                serializedName: "name",
                type: {
                    name: "String"
                }
            },
            message: {
                serializedName: "message",
                type: {
                    name: "String"
                }
            },
            type: {
                serializedName: "type",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var ContainerPropertiesInstanceView = {
    serializedName: "ContainerProperties_instanceView",
    type: {
        name: "Composite",
        className: "ContainerPropertiesInstanceView",
        modelProperties: {
            restartCount: {
                readOnly: true,
                serializedName: "restartCount",
                type: {
                    name: "Number"
                }
            },
            currentState: {
                readOnly: true,
                serializedName: "currentState",
                type: {
                    name: "Composite",
                    className: "ContainerState"
                }
            },
            previousState: {
                readOnly: true,
                serializedName: "previousState",
                type: {
                    name: "Composite",
                    className: "ContainerState"
                }
            },
            events: {
                readOnly: true,
                serializedName: "events",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "Event"
                        }
                    }
                }
            }
        }
    }
};
export var GpuResource = {
    serializedName: "GpuResource",
    type: {
        name: "Composite",
        className: "GpuResource",
        modelProperties: {
            count: {
                required: true,
                serializedName: "count",
                type: {
                    name: "Number"
                }
            },
            sku: {
                required: true,
                serializedName: "sku",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var ResourceRequests = {
    serializedName: "ResourceRequests",
    type: {
        name: "Composite",
        className: "ResourceRequests",
        modelProperties: {
            memoryInGB: {
                required: true,
                serializedName: "memoryInGB",
                type: {
                    name: "Number"
                }
            },
            cpu: {
                required: true,
                serializedName: "cpu",
                type: {
                    name: "Number"
                }
            },
            gpu: {
                serializedName: "gpu",
                type: {
                    name: "Composite",
                    className: "GpuResource"
                }
            }
        }
    }
};
export var ResourceLimits = {
    serializedName: "ResourceLimits",
    type: {
        name: "Composite",
        className: "ResourceLimits",
        modelProperties: {
            memoryInGB: {
                serializedName: "memoryInGB",
                type: {
                    name: "Number"
                }
            },
            cpu: {
                serializedName: "cpu",
                type: {
                    name: "Number"
                }
            },
            gpu: {
                serializedName: "gpu",
                type: {
                    name: "Composite",
                    className: "GpuResource"
                }
            }
        }
    }
};
export var ResourceRequirements = {
    serializedName: "ResourceRequirements",
    type: {
        name: "Composite",
        className: "ResourceRequirements",
        modelProperties: {
            requests: {
                required: true,
                serializedName: "requests",
                type: {
                    name: "Composite",
                    className: "ResourceRequests"
                }
            },
            limits: {
                serializedName: "limits",
                type: {
                    name: "Composite",
                    className: "ResourceLimits"
                }
            }
        }
    }
};
export var VolumeMount = {
    serializedName: "VolumeMount",
    type: {
        name: "Composite",
        className: "VolumeMount",
        modelProperties: {
            name: {
                required: true,
                serializedName: "name",
                type: {
                    name: "String"
                }
            },
            mountPath: {
                required: true,
                serializedName: "mountPath",
                type: {
                    name: "String"
                }
            },
            readOnly: {
                serializedName: "readOnly",
                type: {
                    name: "Boolean"
                }
            }
        }
    }
};
export var ContainerExec = {
    serializedName: "ContainerExec",
    type: {
        name: "Composite",
        className: "ContainerExec",
        modelProperties: {
            command: {
                serializedName: "command",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "String"
                        }
                    }
                }
            }
        }
    }
};
export var ContainerHttpGet = {
    serializedName: "ContainerHttpGet",
    type: {
        name: "Composite",
        className: "ContainerHttpGet",
        modelProperties: {
            path: {
                serializedName: "path",
                type: {
                    name: "String"
                }
            },
            port: {
                required: true,
                serializedName: "port",
                type: {
                    name: "Number"
                }
            },
            scheme: {
                serializedName: "scheme",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var ContainerProbe = {
    serializedName: "ContainerProbe",
    type: {
        name: "Composite",
        className: "ContainerProbe",
        modelProperties: {
            exec: {
                serializedName: "exec",
                type: {
                    name: "Composite",
                    className: "ContainerExec"
                }
            },
            httpGet: {
                serializedName: "httpGet",
                type: {
                    name: "Composite",
                    className: "ContainerHttpGet"
                }
            },
            initialDelaySeconds: {
                serializedName: "initialDelaySeconds",
                type: {
                    name: "Number"
                }
            },
            periodSeconds: {
                serializedName: "periodSeconds",
                type: {
                    name: "Number"
                }
            },
            failureThreshold: {
                serializedName: "failureThreshold",
                type: {
                    name: "Number"
                }
            },
            successThreshold: {
                serializedName: "successThreshold",
                type: {
                    name: "Number"
                }
            },
            timeoutSeconds: {
                serializedName: "timeoutSeconds",
                type: {
                    name: "Number"
                }
            }
        }
    }
};
export var Container = {
    serializedName: "Container",
    type: {
        name: "Composite",
        className: "Container",
        modelProperties: {
            name: {
                required: true,
                serializedName: "name",
                type: {
                    name: "String"
                }
            },
            image: {
                required: true,
                serializedName: "properties.image",
                type: {
                    name: "String"
                }
            },
            command: {
                serializedName: "properties.command",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "String"
                        }
                    }
                }
            },
            ports: {
                serializedName: "properties.ports",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "ContainerPort"
                        }
                    }
                }
            },
            environmentVariables: {
                serializedName: "properties.environmentVariables",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "EnvironmentVariable"
                        }
                    }
                }
            },
            instanceView: {
                readOnly: true,
                serializedName: "properties.instanceView",
                type: {
                    name: "Composite",
                    className: "ContainerPropertiesInstanceView"
                }
            },
            resources: {
                required: true,
                serializedName: "properties.resources",
                type: {
                    name: "Composite",
                    className: "ResourceRequirements"
                }
            },
            volumeMounts: {
                serializedName: "properties.volumeMounts",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "VolumeMount"
                        }
                    }
                }
            },
            livenessProbe: {
                serializedName: "properties.livenessProbe",
                type: {
                    name: "Composite",
                    className: "ContainerProbe"
                }
            },
            readinessProbe: {
                serializedName: "properties.readinessProbe",
                type: {
                    name: "Composite",
                    className: "ContainerProbe"
                }
            }
        }
    }
};
export var AzureFileVolume = {
    serializedName: "AzureFileVolume",
    type: {
        name: "Composite",
        className: "AzureFileVolume",
        modelProperties: {
            shareName: {
                required: true,
                serializedName: "shareName",
                type: {
                    name: "String"
                }
            },
            readOnly: {
                serializedName: "readOnly",
                type: {
                    name: "Boolean"
                }
            },
            storageAccountName: {
                required: true,
                serializedName: "storageAccountName",
                type: {
                    name: "String"
                }
            },
            storageAccountKey: {
                serializedName: "storageAccountKey",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var GitRepoVolume = {
    serializedName: "GitRepoVolume",
    type: {
        name: "Composite",
        className: "GitRepoVolume",
        modelProperties: {
            directory: {
                serializedName: "directory",
                type: {
                    name: "String"
                }
            },
            repository: {
                required: true,
                serializedName: "repository",
                type: {
                    name: "String"
                }
            },
            revision: {
                serializedName: "revision",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var Volume = {
    serializedName: "Volume",
    type: {
        name: "Composite",
        className: "Volume",
        modelProperties: {
            name: {
                required: true,
                serializedName: "name",
                type: {
                    name: "String"
                }
            },
            azureFile: {
                serializedName: "azureFile",
                type: {
                    name: "Composite",
                    className: "AzureFileVolume"
                }
            },
            emptyDir: {
                serializedName: "emptyDir",
                type: {
                    name: "Object"
                }
            },
            secret: {
                serializedName: "secret",
                type: {
                    name: "Dictionary",
                    value: {
                        type: {
                            name: "String"
                        }
                    }
                }
            },
            gitRepo: {
                serializedName: "gitRepo",
                type: {
                    name: "Composite",
                    className: "GitRepoVolume"
                }
            }
        }
    }
};
export var ContainerGroupIdentityUserAssignedIdentitiesValue = {
    serializedName: "ContainerGroupIdentity_userAssignedIdentitiesValue",
    type: {
        name: "Composite",
        className: "ContainerGroupIdentityUserAssignedIdentitiesValue",
        modelProperties: {
            principalId: {
                readOnly: true,
                serializedName: "principalId",
                type: {
                    name: "String"
                }
            },
            clientId: {
                readOnly: true,
                serializedName: "clientId",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var ContainerGroupIdentity = {
    serializedName: "ContainerGroupIdentity",
    type: {
        name: "Composite",
        className: "ContainerGroupIdentity",
        modelProperties: {
            principalId: {
                readOnly: true,
                serializedName: "principalId",
                type: {
                    name: "String"
                }
            },
            tenantId: {
                readOnly: true,
                serializedName: "tenantId",
                type: {
                    name: "String"
                }
            },
            type: {
                serializedName: "type",
                type: {
                    name: "Enum",
                    allowedValues: [
                        "SystemAssigned",
                        "UserAssigned",
                        "SystemAssigned, UserAssigned",
                        "None"
                    ]
                }
            },
            userAssignedIdentities: {
                serializedName: "userAssignedIdentities",
                type: {
                    name: "Dictionary",
                    value: {
                        type: {
                            name: "Composite",
                            className: "ContainerGroupIdentityUserAssignedIdentitiesValue"
                        }
                    }
                }
            }
        }
    }
};
export var ImageRegistryCredential = {
    serializedName: "ImageRegistryCredential",
    type: {
        name: "Composite",
        className: "ImageRegistryCredential",
        modelProperties: {
            server: {
                required: true,
                serializedName: "server",
                type: {
                    name: "String"
                }
            },
            username: {
                required: true,
                serializedName: "username",
                type: {
                    name: "String"
                }
            },
            password: {
                serializedName: "password",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var Port = {
    serializedName: "Port",
    type: {
        name: "Composite",
        className: "Port",
        modelProperties: {
            protocol: {
                serializedName: "protocol",
                type: {
                    name: "String"
                }
            },
            port: {
                required: true,
                serializedName: "port",
                type: {
                    name: "Number"
                }
            }
        }
    }
};
export var IpAddress = {
    serializedName: "IpAddress",
    type: {
        name: "Composite",
        className: "IpAddress",
        modelProperties: {
            ports: {
                required: true,
                serializedName: "ports",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "Port"
                        }
                    }
                }
            },
            type: {
                required: true,
                serializedName: "type",
                type: {
                    name: "String"
                }
            },
            ip: {
                serializedName: "ip",
                type: {
                    name: "String"
                }
            },
            dnsNameLabel: {
                serializedName: "dnsNameLabel",
                type: {
                    name: "String"
                }
            },
            fqdn: {
                readOnly: true,
                serializedName: "fqdn",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var ContainerGroupPropertiesInstanceView = {
    serializedName: "ContainerGroup_properties_instanceView",
    type: {
        name: "Composite",
        className: "ContainerGroupPropertiesInstanceView",
        modelProperties: {
            events: {
                readOnly: true,
                serializedName: "events",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "Event"
                        }
                    }
                }
            },
            state: {
                readOnly: true,
                serializedName: "state",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var LogAnalytics = {
    serializedName: "LogAnalytics",
    type: {
        name: "Composite",
        className: "LogAnalytics",
        modelProperties: {
            workspaceId: {
                required: true,
                serializedName: "workspaceId",
                type: {
                    name: "String"
                }
            },
            workspaceKey: {
                required: true,
                serializedName: "workspaceKey",
                type: {
                    name: "String"
                }
            },
            logType: {
                serializedName: "logType",
                type: {
                    name: "String"
                }
            },
            metadata: {
                serializedName: "metadata",
                type: {
                    name: "Dictionary",
                    value: {
                        type: {
                            name: "String"
                        }
                    }
                }
            }
        }
    }
};
export var ContainerGroupDiagnostics = {
    serializedName: "ContainerGroupDiagnostics",
    type: {
        name: "Composite",
        className: "ContainerGroupDiagnostics",
        modelProperties: {
            logAnalytics: {
                serializedName: "logAnalytics",
                type: {
                    name: "Composite",
                    className: "LogAnalytics"
                }
            }
        }
    }
};
export var ContainerGroupNetworkProfile = {
    serializedName: "ContainerGroupNetworkProfile",
    type: {
        name: "Composite",
        className: "ContainerGroupNetworkProfile",
        modelProperties: {
            id: {
                required: true,
                serializedName: "id",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var DnsConfiguration = {
    serializedName: "DnsConfiguration",
    type: {
        name: "Composite",
        className: "DnsConfiguration",
        modelProperties: {
            nameServers: {
                required: true,
                serializedName: "nameServers",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "String"
                        }
                    }
                }
            },
            searchDomains: {
                serializedName: "searchDomains",
                type: {
                    name: "String"
                }
            },
            options: {
                serializedName: "options",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var Resource = {
    serializedName: "Resource",
    type: {
        name: "Composite",
        className: "Resource",
        modelProperties: {
            id: {
                readOnly: true,
                serializedName: "id",
                type: {
                    name: "String"
                }
            },
            name: {
                readOnly: true,
                serializedName: "name",
                type: {
                    name: "String"
                }
            },
            type: {
                readOnly: true,
                serializedName: "type",
                type: {
                    name: "String"
                }
            },
            location: {
                serializedName: "location",
                type: {
                    name: "String"
                }
            },
            tags: {
                serializedName: "tags",
                type: {
                    name: "Dictionary",
                    value: {
                        type: {
                            name: "String"
                        }
                    }
                }
            }
        }
    }
};
export var ContainerGroup = {
    serializedName: "ContainerGroup",
    type: {
        name: "Composite",
        className: "ContainerGroup",
        modelProperties: tslib_1.__assign({}, Resource.type.modelProperties, { identity: {
                serializedName: "identity",
                type: {
                    name: "Composite",
                    className: "ContainerGroupIdentity"
                }
            }, provisioningState: {
                readOnly: true,
                serializedName: "properties.provisioningState",
                type: {
                    name: "String"
                }
            }, containers: {
                required: true,
                serializedName: "properties.containers",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "Container"
                        }
                    }
                }
            }, imageRegistryCredentials: {
                serializedName: "properties.imageRegistryCredentials",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "ImageRegistryCredential"
                        }
                    }
                }
            }, restartPolicy: {
                serializedName: "properties.restartPolicy",
                type: {
                    name: "String"
                }
            }, ipAddress: {
                serializedName: "properties.ipAddress",
                type: {
                    name: "Composite",
                    className: "IpAddress"
                }
            }, osType: {
                required: true,
                serializedName: "properties.osType",
                type: {
                    name: "String"
                }
            }, volumes: {
                serializedName: "properties.volumes",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "Volume"
                        }
                    }
                }
            }, instanceView: {
                readOnly: true,
                serializedName: "properties.instanceView",
                type: {
                    name: "Composite",
                    className: "ContainerGroupPropertiesInstanceView"
                }
            }, diagnostics: {
                serializedName: "properties.diagnostics",
                type: {
                    name: "Composite",
                    className: "ContainerGroupDiagnostics"
                }
            }, networkProfile: {
                serializedName: "properties.networkProfile",
                type: {
                    name: "Composite",
                    className: "ContainerGroupNetworkProfile"
                }
            }, dnsConfig: {
                serializedName: "properties.dnsConfig",
                type: {
                    name: "Composite",
                    className: "DnsConfiguration"
                }
            } })
    }
};
export var OperationDisplay = {
    serializedName: "Operation_display",
    type: {
        name: "Composite",
        className: "OperationDisplay",
        modelProperties: {
            provider: {
                serializedName: "provider",
                type: {
                    name: "String"
                }
            },
            resource: {
                serializedName: "resource",
                type: {
                    name: "String"
                }
            },
            operation: {
                serializedName: "operation",
                type: {
                    name: "String"
                }
            },
            description: {
                serializedName: "description",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var Operation = {
    serializedName: "Operation",
    type: {
        name: "Composite",
        className: "Operation",
        modelProperties: {
            name: {
                required: true,
                serializedName: "name",
                type: {
                    name: "String"
                }
            },
            display: {
                required: true,
                serializedName: "display",
                type: {
                    name: "Composite",
                    className: "OperationDisplay"
                }
            },
            origin: {
                serializedName: "origin",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var OperationListResult = {
    serializedName: "OperationListResult",
    type: {
        name: "Composite",
        className: "OperationListResult",
        modelProperties: {
            value: {
                serializedName: "value",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "Operation"
                        }
                    }
                }
            },
            nextLink: {
                serializedName: "nextLink",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var UsageName = {
    serializedName: "Usage_name",
    type: {
        name: "Composite",
        className: "UsageName",
        modelProperties: {
            value: {
                readOnly: true,
                serializedName: "value",
                type: {
                    name: "String"
                }
            },
            localizedValue: {
                readOnly: true,
                serializedName: "localizedValue",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var Usage = {
    serializedName: "Usage",
    type: {
        name: "Composite",
        className: "Usage",
        modelProperties: {
            unit: {
                readOnly: true,
                serializedName: "unit",
                type: {
                    name: "String"
                }
            },
            currentValue: {
                readOnly: true,
                serializedName: "currentValue",
                type: {
                    name: "Number"
                }
            },
            limit: {
                readOnly: true,
                serializedName: "limit",
                type: {
                    name: "Number"
                }
            },
            name: {
                readOnly: true,
                serializedName: "name",
                type: {
                    name: "Composite",
                    className: "UsageName"
                }
            }
        }
    }
};
export var UsageListResult = {
    serializedName: "UsageListResult",
    type: {
        name: "Composite",
        className: "UsageListResult",
        modelProperties: {
            value: {
                readOnly: true,
                serializedName: "value",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "Usage"
                        }
                    }
                }
            }
        }
    }
};
export var Logs = {
    serializedName: "Logs",
    type: {
        name: "Composite",
        className: "Logs",
        modelProperties: {
            content: {
                serializedName: "content",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var ContainerExecRequestTerminalSize = {
    serializedName: "ContainerExecRequest_terminalSize",
    type: {
        name: "Composite",
        className: "ContainerExecRequestTerminalSize",
        modelProperties: {
            rows: {
                serializedName: "rows",
                type: {
                    name: "Number"
                }
            },
            cols: {
                serializedName: "cols",
                type: {
                    name: "Number"
                }
            }
        }
    }
};
export var ContainerExecRequest = {
    serializedName: "ContainerExecRequest",
    type: {
        name: "Composite",
        className: "ContainerExecRequest",
        modelProperties: {
            command: {
                serializedName: "command",
                type: {
                    name: "String"
                }
            },
            terminalSize: {
                serializedName: "terminalSize",
                type: {
                    name: "Composite",
                    className: "ContainerExecRequestTerminalSize"
                }
            }
        }
    }
};
export var ContainerExecResponse = {
    serializedName: "ContainerExecResponse",
    type: {
        name: "Composite",
        className: "ContainerExecResponse",
        modelProperties: {
            webSocketUri: {
                serializedName: "webSocketUri",
                type: {
                    name: "String"
                }
            },
            password: {
                serializedName: "password",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var ContainerGroupListResult = {
    serializedName: "ContainerGroupListResult",
    type: {
        name: "Composite",
        className: "ContainerGroupListResult",
        modelProperties: {
            value: {
                serializedName: "",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "ContainerGroup"
                        }
                    }
                }
            },
            nextLink: {
                serializedName: "nextLink",
                type: {
                    name: "String"
                }
            }
        }
    }
};
