// @azure/arm-containerinstance/esm/operations/serviceAssociationLink.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
import * as msRest from "@azure/ms-rest-js";
import * as Mappers from "../models/serviceAssociationLinkMappers";
import * as Parameters from "../models/parameters";
/** Class representing a ServiceAssociationLink. */
var ServiceAssociationLink = /** @class */ (function () {
    /**
     * Create a ServiceAssociationLink.
     * @param {ContainerInstanceManagementClientContext} client Reference to the service client.
     */
    function ServiceAssociationLink(client) {
        this.client = client;
    }
    ServiceAssociationLink.prototype.deleteMethod = function (resourceGroupName, virtualNetworkName, subnetName, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            virtualNetworkName: virtualNetworkName,
            subnetName: subnetName,
            options: options
        }, deleteMethodOperationSpec, callback);
    };
    return ServiceAssociationLink;
}());
export { ServiceAssociationLink };
// Operation Specifications
var serializer = new msRest.Serializer(Mappers);
var deleteMethodOperationSpec = {
    httpMethod: "DELETE",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/virtualNetworks/{virtualNetworkName}/subnets/{subnetName}/providers/Microsoft.ContainerInstance/serviceAssociationLinks/default",
    urlParameters: [
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.virtualNetworkName,
        Parameters.subnetName
    ],
    queryParameters: [
        Parameters.apiVersion
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    responses: {
        200: {},
        204: {},
        default: {
            bodyMapper: Mappers.CloudError
        }
    },
    serializer: serializer
};
